core/embed/rust/src/ui/layout/util.rs

```diff
@@ -32,3 +35,98 @@ where
     // Returns error if array.len() != N
     vec.into_array().map_err(|_| err)
 }
+
+fn objslice_at<'a>(
+    objslice: &'a [Paragraph<Obj>],
+    index: usize,
+    offset: usize,
+    buffer: &'a mut [u8],
+) -> Paragraph<&'a str> {
+    const HEX_LOWER: [u8; 16] = *b"0123456789abcdef";
+    let par = &objslice[index];
+
+    // Handle None as empty string.
+    if par.content() == Obj::const_none() {
+        return par.with_content("");
+    }
+
+    // Handle str.
+    if StrBuffer::try_from(par.content()).is_ok() {
```

Comment:
why not `if let strbuf = StrBuffer()...` and then `strbuf.as_ref()` instead of doing the get_buffer unsafe dance here?

Reply:
```
error[E0515]: cannot return value referencing local variable `strbuf`
  --> src/ui/layout/util.rs:60:16
   |
59 |         let s = strbuf.as_ref();
   |                 --------------- `strbuf` is borrowed here
60 |         return par.with_content(&s[offset..]);
   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ returns a value referencing data owned by the current function

```

Reply:
OK not really
```
error[E0106]: missing lifetime specifier
  --> src/micropython/buffer.rs:58:26
   |
58 |     fn into_ref(self) -> &str {
   |                          ^ expected named lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from
```

Reply:
But this works, not sure if good idea though.
```rust
    pub fn into_ref<'a>(self) -> &'a str {
        if self.ptr.is_null() {
            return ""
        }
        let slice = unsafe { slice::from_raw_parts(self.ptr, self.len) };
        unsafe { str::from_utf8_unchecked(slice) }
    }

```

Reply:
an alternative is to [teach `StrBuffer` about slicing](https://stackoverflow.com/a/57203324) while keeping the original str pointer, and make the inner Paragraphs based on `StrBuffer`s

Reply:
What if we wrapped this in a function that ties the lifetime of the result to the source `Obj` (needs `Paragraph::content` to return a reference)? Is it useless because `Obj` implements `Copy`?
```rust
pub fn obj_as_str(obj: &Obj) -> Result<&str, Error> {
    let strbuf = StrBuffer::try_from(*obj)?;
    Ok(strbuf.into_ref())
}
```

Reply:
> Is it useless because `Obj` implements `Copy`?

I think the general problem is not the `Copy`, but that `Obj` is a pointer and not an owner.
*But* what the code above seems to say is, as long as I hold _the pointer_, i.e., a piece of memory with an address of the uPy object, then the `&str` slice is valid.
Which seems to be exactly true?

One potential risk is that the obj in question is once removed from the actual string data, i.e., in theory, someone could mutate contents of the obj to point to another string. I don't think uPy allows that though.

Still, it feels like cheating, so I want @jpochyla to review it first.

Reply:
> in theory, someone could mutate contents of the obj to point to another string

even more specifically, since `Obj` is `Copy`, someone could try to grab another copy mutably from Rust. I am like 80 % sure that can't be done to a string either -- in general, the fact that this `Obj` is a string type is giving us a lot to work with.

Reply:
Yeah the last suggestion seems like a best bet. The `into_ref` thing cannot really be a safe fn under the general safety guidelines (it's unsound -- possible to use unsafely from a safe code.) I don't really have time to make an authoritative opinion, though.

Reply:
Proposed solution in [0cc990d](https://github.com/trezor/trezor-firmware/pull/2575/commits/0cc990dd67e1b8b444a7e0a9a7d5992dfea7382d).